File: src/subscription.py
from __future__ import annotations

import base64
import logging
import re
from typing import List, Sequence, Tuple
from urllib.parse import unquote

import requests

logger = logging.getLogger(__name__)

SUPPORTED_SCHEMES = (
    "vmess://",
    "vless://",
    "trojan://",
    "ss://",
    "ssr://",
    "hysteria2://",
    "hy2://",
    "tuic://",
    "wireguard://",
)

DEFAULT_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/120.0.0.0 Safari/537.36"
    ),
    "Accept": "*/*",
}

_SCHEME_RE = re.compile(
    r"(?i)((?:vmess|vless|trojan|ss|ssr|hysteria2|hy2|tuic|wireguard)://\S+)"
)


def _try_b64_decode(data: str) -> str | None:
    cleaned = "".join(data.split())
    if not cleaned:
        return None

    for decoder in (base64.b64decode, base64.urlsafe_b64decode):
        for pad in ("", "=", "==", "==="):
            try:
                raw = decoder(cleaned + pad, validate=False)
                text = raw.decode("utf-8", errors="ignore")
                if text.strip():
                    return text
            except Exception:
                continue
    return None


def decode_subscription_body(body: str) -> str:
    body = body.strip()
    if not body:
        return ""

    lower = body.lower()
    if any(scheme in lower for scheme in SUPPORTED_SCHEMES):
        return body

    decoded = _try_b64_decode(body)
    return decoded if decoded is not None else body


def _split_share_links(line: str) -> List[str]:
    matches = _SCHEME_RE.findall(line)
    if matches:
        return [m.rstrip(",;") for m in matches]

    lower = line.lower()
    for scheme in SUPPORTED_SCHEMES:
        if lower.startswith(scheme):
            return [line]
    return []


def extract_configs(text: str) -> List[str]:
    configs: List[str] = []
    seen: set[str] = set()

    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or line.startswith("//"):
            continue

        if "%" in line and "://" not in line[:12]:
            try:
                line = unquote(line).strip()
            except Exception:
                pass

        for link in _split_share_links(line):
            link = link.strip().rstrip("\r")
            if not link or link in seen:
                continue
            seen.add(link)
            configs.append(link)

    return configs


def fetch_subscription(
    url: str,
    timeout: float = 30.0,
    session: requests.Session | None = None,
) -> Tuple[List[str], str | None]:
    url = url.strip()
    if not url:
        return [], "empty url"

    sess = session or requests.Session()
    try:
        resp = sess.get(url, headers=DEFAULT_HEADERS, timeout=timeout)
        resp.raise_for_status()
        try:
            body = resp.content.decode("utf-8", errors="ignore")
        except Exception:
            body = resp.text

        configs = extract_configs(decode_subscription_body(body))
        if not configs:
            return [], "no proxy links found"
        return configs, None
    except requests.Timeout:
        return [], "timeout"
    except requests.RequestException as exc:
        return [], "fetch failed: " + str(exc)
    except Exception as exc:
        return [], str(exc)


def fetch_all_subscriptions(
    urls: Sequence[str],
    timeout: float = 30.0,
    progress_cb=None,
) -> Tuple[List[str], List[str]]:
    all_configs: List[str] = []
    seen: set[str] = set()
    errors: List[str] = []

    raw_links: List[str] = []
    real_urls: List[str] = []
    for item in urls:
        item = item.strip()
        if not item:
            continue
        lower = item.lower()
        if any(lower.startswith(s) for s in SUPPORTED_SCHEMES):
            raw_links.append(item)
        else:
            real_urls.append(item)

    for link in raw_links:
        if link not in seen:
            seen.add(link)
            all_configs.append(link)

    session = requests.Session()
    total = len(real_urls)
    for idx, url in enumerate(real_urls, start=1):
        if progress_cb:
            progress_cb(idx, total, url)
        configs, err = fetch_subscription(url, timeout=timeout, session=session)
        if err:
            errors.append(url + ": " + err)
            logger.warning("subscription error for %s: %s", url, err)
            continue
        for cfg in configs:
            if cfg not in seen:
                seen.add(cfg)
                all_configs.append(cfg)

    return all_configs, errors


def parse_url_list(text: str) -> List[str]:
    if not text or not text.strip():
        return []

    parts: List[str] = []
    for chunk in re.split(r"[\n,;]+", text):
        chunk = chunk.strip()
        if chunk:
            parts.append(chunk)
    return parts

File: app/lib/core/subscription.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

const List<String> supportedSchemes = <String>[
  'vmess://',
  'vless://',
  'trojan://',
  'ss://',
  'ssr://',
  'hysteria2://',
  'hy2://',
  'tuic://',
  'wireguard://',
];

final RegExp _linkPattern = RegExp(
  r'(?:vmess|vless|trojan|ss|ssr|hysteria2|hy2|tuic|wireguard):\/\/\S+',
  caseSensitive: false,
);

class SourceResult {
  SourceResult({
    required this.url,
    required this.links,
    this.error = '',
  });

  final String url;
  final List<String> links;
  final String error;

  bool get ok => error.isEmpty;
}

String? _tryBase64(String data) {
  final String cleaned = data.replaceAll(RegExp(r'\s'), '');
  if (cleaned.isEmpty) {
    return null;
  }
  for (final Codec<List<int>, String> codec in <Codec<List<int>, String>>[
    base64,
    base64Url,
  ]) {
    for (int pad = 0; pad < 4; pad++) {
      try {
        final String padded = cleaned + '=' * pad;
        final String text = utf8.decode(
          codec.decode(padded),
          allowMalformed: true,
        );
        if (text.trim().isNotEmpty) {
          return text;
        }
      } catch (_) {
        continue;
      }
    }
  }
  return null;
}

String decodeBody(String body) {
  final String trimmed = body.trim();
  if (trimmed.isEmpty) {
    return '';
  }
  final String lower = trimmed.toLowerCase();
  for (final String scheme in supportedSchemes) {
    if (lower.contains(scheme)) {
      return trimmed;
    }
  }
  return _tryBase64(trimmed) ?? trimmed;
}

List<String> extractLinks(String text) {
  final List<String> links = <String>[];
  final Set<String> seen = <String>{};

  for (final String rawLine in text.split('\n')) {
    String line = rawLine.trim();
    if (line.isEmpty || line.startsWith('#') || line.startsWith('//')) {
      continue;
    }
    final int head = line.length < 12 ? line.length : 12;
    if (line.contains('%') && !line.substring(0, head).contains('://')) {
      try {
        line = Uri.decodeFull(line).trim();
      } catch (_) {
        line = rawLine.trim();
      }
    }
    for (final RegExpMatch match in _linkPattern.allMatches(line)) {
      String link = match.group(0) ?? '';
      link = link.trim().replaceAll(RegExp(r'[,;]+$'), '');
      if (link.isEmpty || !seen.add(link)) {
        continue;
      }
      links.add(link);
    }
  }

  return links;
}

List<String> splitSourceList(String text) {
  final List<String> parts = <String>[];
  for (final String chunk in text.split(RegExp(r'[\n,;]+'))) {
    final String trimmed = chunk.trim();
    if (trimmed.isNotEmpty) {
      parts.add(trimmed);
    }
  }
  return parts;
}

class SubscriptionFetcher {
  SubscriptionFetcher({
    this.timeout = const Duration(seconds: 25),
    this.allowInvalidCertificates = false,
  });

  final Duration timeout;
  final bool allowInvalidCertificates;

  static const String _userAgent =
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 '
      '(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';

  HttpClient _client() {
    final HttpClient client = HttpClient()
      ..connectionTimeout = timeout
      ..idleTimeout = const Duration(seconds: 5)
      ..userAgent = _userAgent;
    if (allowInvalidCertificates) {
      client.badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
    }
    return client;
  }

  Future<SourceResult> fetchOne(String url) async {
    final String trimmed = url.trim();
    if (trimmed.isEmpty) {
      return SourceResult(url: url, links: <String>[], error: 'empty url');
    }

    final String lower = trimmed.toLowerCase();
    for (final String scheme in supportedSchemes) {
      if (lower.startsWith(scheme)) {
        return SourceResult(url: trimmed, links: <String>[trimmed]);
      }
    }

    final Uri? uri = Uri.tryParse(trimmed);
    if (uri == null || !uri.hasScheme || uri.host.isEmpty) {
      return SourceResult(url: trimmed, links: <String>[], error: 'invalid url');
    }

    final HttpClient client = _client();
    try {
      final HttpClientRequest request =
          await client.getUrl(uri).timeout(timeout);
      request.followRedirects = true;
      request.maxRedirects = 5;
      request.headers.set(HttpHeaders.acceptHeader, '*/*');
      final HttpClientResponse response = await request.close().timeout(timeout);
      if (response.statusCode < 200 || response.statusCode >= 300) {
        return SourceResult(
          url: trimmed,
          links: <String>[],
          error: 'http ${response.statusCode}',
        );
      }
      final List<int> bytes = await _collect(response).timeout(timeout);
      final String body = utf8.decode(bytes, allowMalformed: true);
      final List<String> links = extractLinks(decodeBody(body));
      if (links.isEmpty) {
        return SourceResult(
          url: trimmed,
          links: <String>[],
          error: 'no links found',
        );
      }
      return SourceResult(url: trimmed, links: links);
    } on TimeoutException {
      return SourceResult(url: trimmed, links: <String>[], error: 'timeout');
    } catch (error) {
      return SourceResult(
        url: trimmed,
        links: <String>[],
        error: _shortError(error),
      );
    } finally {
      client.close(force: true);
    }
  }

  Future<List<String>> fetchAll(
    List<String> urls, {
    int concurrency = 8,
    void Function(int done, int total, SourceResult result)? onSource,
  }) async {
    final List<String> merged = <String>[];
    final Set<String> seen = <String>{};
    int done = 0;

    final List<String> queue = List<String>.from(urls);
    final int workers = concurrency < 1 ? 1 : concurrency;

    Future<void> worker() async {
      while (true) {
        String? next;
        if (queue.isNotEmpty) {
          next = queue.removeAt(0);
        }
        if (next == null) {
          return;
        }
        final SourceResult result = await fetchOne(next);
        done += 1;
        for (final String link in result.links) {
          if (seen.add(link)) {
            merged.add(link);
          }
        }
        if (onSource != null) {
          onSource(done, urls.length, result);
        }
      }
    }

    await Future.wait(<Future<void>>[
      for (int i = 0; i < workers; i++) worker(),
    ]);

    return merged;
  }

  static Future<List<int>> _collect(Stream<List<int>> stream) async {
    final List<int> bytes = <int>[];
    await for (final List<int> chunk in stream) {
      bytes.addAll(chunk);
      if (bytes.length > 12 * 1024 * 1024) {
        break;
      }
    }
    return bytes;
  }

  static String _shortError(Object error) {
    final String text = error.toString();
    if (text.length <= 90) {
      return text;
    }
    return '${text.substring(0, 90)}...';
  }
}
